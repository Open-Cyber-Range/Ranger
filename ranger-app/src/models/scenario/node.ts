import {type Source} from './common';

export enum NodeType {
  VM,
  Switch,
}

export type Resources = {
  ram: number;
  cpu: number;
};

export type Node = {
  type_field: NodeType;
  description?: string;
  resources?: Resources;
  source?: Source;
  features?: Record<string, string>;
  conditions?: Record<string, string>;
  vulnerabilities?: string[];
  roles?: Record<string, string>;
};

export type Nodes = Record<string, Node>;
