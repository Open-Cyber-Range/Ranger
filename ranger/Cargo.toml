[package]
name = "ranger"
version = "2.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[package.metadata.deb]
name = "ranger-server"
maintainer = "C14 developers <developers@cr14.ee>"
copyright = "MIT"
depends = "$auto, systemd"
section = "utility"
priority = "optional"
changelog = "changelog"
assets = [
  [
    "target/release/ranger",
    "/var/opt/ranger/bin/",
    "755",
  ],
  [
    "assets/default-production-config.yml",
    "/etc/opt/ranger/ranger/config.yml",
    "644",
  ],
  [
    "../ranger-app/build/assets/*",
    "/etc/opt/ranger/ranger-app/assets",
    "644",
  ],
  [
    "../ranger-app/build/*",
    "/etc/opt/ranger/ranger-app/",
    "644",
  ],
]
systemd-units = { unit-name = "ranger", unit-scripts = "assets/unit-scripts" }

[dependencies]
actix = "0.13"
actix-multipart = "0.6"
actix-web = "4"
actix-web-actors = "4"
actix-web-grants = "3"
actix-http = { version = "3", features = ["http2", "ws"] }
actix-web-httpauth = "0.8"
anyhow = "1"
async-trait = "0.1"
bigdecimal = { version = "0.4", features = ["serde"] }
chrono = { version = "0.4", features = ["serde"] }
diesel = { version = "2.1", features = ["mysql", "r2d2", "chrono", "numeric"] }
diesel_migrations = "2.1"
env_logger = "0.9"
futures = "0.3"
futures-util = "0.3"
jsonwebtoken = "8"
keycloak = { git = "https://github.com/ebakoba/keycloak", branch = "master" }
log = "0.4"
lazy_static = "1.4"
lettre = { version = "0.10", default-features = false, features = [
  "smtp-transport",
  "pool",
  "hostname",
  "builder",
  "rustls",
  "rustls-tls",
] }
sha3 = "0.10"
digest = "0.10"
rand = "0.8"
ranger-grpc = { path = "../ranger-grpc" }
reqwest = { version = "0.11", default-features = false, features = [
  "json",
  "rustls",
  "rustls-tls",
  "tokio-rustls",
  "hyper-rustls",
  "rustls-native-certs",
] }
sdl-parser = "0"
serde = { version = "1.0", features = ["derive"] }
serde_json = "1"
serde_yaml = "0.9"
semver = "1"
tempfile = "3"
thiserror = "1"
toml-query = "0.10"
toml = "0.5.6"
tokio = { version = "1", features = [
  "rt-multi-thread",
  "time",
  "fs",
  "macros",
  "net",
  "rt",
] }
tonic = "0.8"
uuid = { version = "1.1", features = ["v4", "serde", "fast-rng"] }

[dev-dependencies]
actix-rt = "2"
insta = { version = "1", features = ["redactions", "yaml"] }
sdl-parser = { version = "0", features = ["test"] }
