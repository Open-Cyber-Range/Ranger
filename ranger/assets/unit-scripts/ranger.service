[Unit]
Description=ranger
After=network.target

[Service]
ExecStart=/var/opt/ranger/bin/ranger /etc/opt/ranger/ranger/config.yml
Type=simple
Restart=on-failure
RestartSec=10
StartLimitInterval=10m
StartLimitBurst=5

[Install]
WantedBy=multi-user.target
