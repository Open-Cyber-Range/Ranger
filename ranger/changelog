ranger (0.2.0) stretch; urgency=medium

  * Initial Release
  * Adds web application support

 -- Kaarel Allemann <kaarel.allemann@cr14.ee>  Sat, 05 Nov 2022 19:23:51 +0200
