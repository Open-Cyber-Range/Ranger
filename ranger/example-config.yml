host: localhost
port: 8080
repository:
  folder: /tmp/test-repo
  username: some-username
  email: some@email.com
package_folder: /tmp/packages
