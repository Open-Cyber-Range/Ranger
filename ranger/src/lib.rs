pub(crate) mod configuration;
mod constants;
pub mod errors;
pub mod models;
pub mod routes;
pub(crate) mod schema;
pub(crate) mod services;
pub(crate) mod utilities;

use crate::services::database::Database;
use actix::{Actor, Addr};
use anyhow::Result;
use configuration::{read_configuration, Configuration};
use log::{error, info};
use services::{
    deployer::{DeployerDistribution, DeployerFactory},
    deployment::DeploymentManager,
    mailer::Mailer,
    scheduler::Scheduler,
    websocket::WebSocketManager,
};

pub struct AppState {
    pub database_address: Addr<Database>,
    pub deployment_manager_address: Addr<DeploymentManager>,
    pub configuration: Configuration,
    pub websocket_manager_address: Addr<WebSocketManager>,
}

impl AppState {
    pub fn new(
        configuration: &configuration::Configuration,
        distributor: &Addr<DeployerDistribution>,
        database: &Addr<Database>,
        websocket_manager: &Addr<WebSocketManager>,
    ) -> Self {
        let schduler_address = Scheduler::new().start();
        AppState {
            database_address: database.clone(),
            deployment_manager_address: DeploymentManager::new(
                schduler_address,
                distributor.clone(),
                database.clone(),
                configuration.deployment_groups.clone(),
                configuration.default_deployment_group.to_string(),
            )
            .start(),
            configuration: configuration.to_owned(),
            websocket_manager_address: websocket_manager.clone(),
        }
    }
}

pub async fn app_setup(environment_arguments: Vec<String>) -> Result<(String, u16, AppState)> {
    let configuration = read_configuration(environment_arguments)?;
    let deployer_factory = DeployerFactory::new(&configuration.deployers)
        .await
        .unwrap_or_else(|error| panic!("Failed to create deployer distribution: {error}"))
        .start();

    let deployer_distributor = DeployerDistribution::new(
        deployer_factory,
        configuration.deployers.keys().cloned().collect(),
    )
    .await
    .unwrap_or_else(|error| panic!("Failed to create deployer distribution: {error}"))
    .start();

    let websocket_manager = WebSocketManager::new().start();
    let database = Database::try_new(&configuration.database_url, &websocket_manager)
        .unwrap_or_else(|error| {
            panic!(
                "Failed to create database connection to {} due to: {error}",
                &configuration.database_url
            )
        })
        .start();

    send_mail(configuration.clone());

    let app_state = AppState::new(
        &configuration,
        &deployer_distributor,
        &database,
        &websocket_manager,
    );
    Ok((configuration.host, configuration.port, app_state))
}

pub fn send_mail(configuration: Configuration) {
    if let Some(mailer_configuration) = configuration.mailer_configuration {
        match Mailer::send_mail(
            mailer_configuration,
            "real.person@real-email.com".to_string(),
        ) {
            Ok(_) => info!("Mail sent successfully!"),
            Err(e) => error!("Mailer failed: {:?}", e),
        }
    }
}
