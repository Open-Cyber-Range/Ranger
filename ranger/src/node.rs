use actix::prelude::*;
use actix::{Actor, Handler, Message};
use anyhow::{Ok, Result};
use ranger_grpc::{node_service_client::NodeServiceClient, Identifier, Node};
use tonic::transport::Channel;

#[derive(Message)]
#[rtype(result = "Result<Identifier, anyhow::Error>")]
pub struct CreateNode(pub Node);

#[derive(Message)]
#[rtype(result = "Result<(), anyhow::Error>")]
pub struct DeleteNode(pub String);

pub struct NodeClient {
    client: NodeServiceClient<Channel>,
}

impl NodeClient {
    pub async fn new(server_address: String) -> Result<Self> {
        Ok(Self {
            client: NodeServiceClient::connect(server_address).await?,
        })
    }
}

impl Actor for NodeClient {
    type Context = actix::Context<Self>;
}

impl Handler<CreateNode> for NodeClient {
    type Result = ResponseFuture<Result<Identifier>>;

    fn handle(&mut self, msg: CreateNode, _ctx: &mut Self::Context) -> Self::Result {
        let node = msg.0;
        let mut client = self.client.clone();
        Box::pin(async move {
            let result = client.create(tonic::Request::new(node)).await;
            if let Err(status) = result {
                return Err(anyhow::anyhow!("{:?}", status));
            }
            Ok(result?.into_inner())
        })
    }
}

impl Handler<DeleteNode> for NodeClient {
    type Result = ResponseFuture<Result<()>>;

    fn handle(&mut self, msg: DeleteNode, _ctx: &mut Self::Context) -> Self::Result {
        let node_id = msg.0;
        let mut client = self.client.clone();
        Box::pin(async move {
            let result = client
                .delete(tonic::Request::new(Identifier { value: node_id }))
                .await;
            if let Err(status) = result {
                return Err(anyhow::anyhow!("{:?}", status));
            }
            Ok(())
        })
    }
}
