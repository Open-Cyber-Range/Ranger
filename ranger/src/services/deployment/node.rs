use super::condition::DeployableConditions;
use super::feature::DeployableFeatures;
use crate::models::helpers::{deployer_type::DeployerType, uuid::Uuid};
use crate::models::{Deployment, DeploymentElement, ElementStatus, Exercise};
use crate::services::client::{Deployable, DeploymentInfo};
use crate::services::database::deployment::{
    CreateDeploymentElement, GetDeploymentElementByDeploymentIdByScenarioReference,
    UpdateDeploymentElement,
};
use crate::services::database::Database;
use crate::services::deployer::{Deploy, DeployerDistribution, UnDeploy};
use crate::services::scheduler::{CreateDeploymentSchedule, Scheduler};
use actix::Addr;
use anyhow::{anyhow, Ok, Result};
use async_trait::async_trait;
use futures::future::try_join_all;
use ranger_grpc::{
    capabilities::DeployerTypes, Configuration, DeploySwitch, DeployVirtualMachine, Identifier,
    MetaInfo, Switch, VirtualMachine,
};
use sdl_parser::{
    common::Source,
    node::{Node, NodeType},
    Scenario,
};

impl Deployable
    for (
        String,
        Node,
        Deployment,
        String,
        Vec<String>,
        Option<String>,
    )
{
    fn try_to_deployment_command(&self) -> Result<Box<dyn DeploymentInfo>> {
        let (name, node, deployment, exercise_name, links, template_id) = self;
        let meta_info = Some(MetaInfo {
            exercise_name: exercise_name.to_string(),
            deployment_name: deployment.name.to_string(),
        });
        let name = name.to_string();
        let links = links.to_vec();
        Ok(match node.type_field {
            NodeType::VM => Box::new(DeployVirtualMachine {
                virtual_machine: Some(VirtualMachine {
                    name,
                    links,
                    configuration: node.resources.as_ref().map(|resource| Configuration {
                        cpu: resource.cpu,
                        ram: resource.ram,
                    }),
                    template_id: template_id
                        .clone()
                        .ok_or_else(|| anyhow!("Template not found"))?,
                }),
                meta_info,
            }),
            NodeType::Switch => Box::new(DeploySwitch {
                switch: Some(Switch { name, links }),
                meta_info,
            }),
        })
    }
}

#[async_trait]
pub trait DeployableNodes {
    async fn deploy_nodes(
        &self,
        distributor_address: &Addr<DeployerDistribution>,
        scheduler_address: &Addr<Scheduler>,
        database_address: &Addr<Database>,
        exercise: &Exercise,
        deployment: &Deployment,
        deployers: &[String],
    ) -> Result<()>;
}

async fn get_template_id(
    deployment_id: Uuid,
    source: &Option<Source>,
    database_address: &Addr<Database>,
) -> Result<Option<String>> {
    if let Some(source) = source {
        let deployment_element = database_address
            .send(GetDeploymentElementByDeploymentIdByScenarioReference(
                deployment_id,
                Box::new(source.to_owned()),
                true,
            ))
            .await??;

        return Ok(deployment_element.handler_reference);
    }
    Ok(None)
}

#[async_trait]
impl DeployableNodes for Scenario {
    async fn deploy_nodes(
        &self,
        distributor_address: &Addr<DeployerDistribution>,
        scheduler_address: &Addr<Scheduler>,
        database_address: &Addr<Database>,
        exercise: &Exercise,
        deployment: &Deployment,
        deployers: &[String],
    ) -> Result<()> {
        let deployment_schedule = scheduler_address
            .send(CreateDeploymentSchedule(self.clone()))
            .await??;
        for tranche in deployment_schedule.iter() {
            try_join_all(
                tranche
                    .iter()
                    .map(|(unique_name, node, infra_node)| async move {
                        let deployer_type = match node.type_field {
                            sdl_parser::node::NodeType::VM => DeployerTypes::VirtualMachine,
                            sdl_parser::node::NodeType::Switch => DeployerTypes::Switch,
                        };
                        let mut deployment_element = database_address
                            .send(CreateDeploymentElement(
                                exercise.id,
                                DeploymentElement::new_ongoing(
                                    deployment.id,
                                    Box::new(unique_name.to_string()),
                                    deployer_type,
                                ),
                                true,
                            ))
                            .await??;
                        let links =
                            try_join_all(infra_node.links.clone().unwrap_or_default().iter().map(
                                |link_name| async move {
                                    let deployment_element = database_address
                                        .send(
                                            GetDeploymentElementByDeploymentIdByScenarioReference(
                                                deployment.id,
                                                Box::new(link_name.to_string()),
                                                true,
                                            ),
                                        )
                                        .await??;
                                    deployment_element
                                        .handler_reference
                                        .ok_or_else(|| anyhow!("Handler reference not found"))
                                },
                            ))
                            .await?;
                        let template_id =
                            get_template_id(deployment.id, &node.source, database_address).await?;
                        let command = (
                            unique_name.clone(),
                            node.clone(),
                            deployment.to_owned(),
                            exercise.name.to_string(),
                            links,
                            template_id.clone(),
                        )
                            .try_to_deployment_command()?;

                        match distributor_address
                            .send(Deploy(deployer_type, command, deployers.to_owned()))
                            .await?
                        {
                            anyhow::Result::Ok(client_response) => {
                                let id = Identifier::try_from(client_response)?.value;

                                deployment_element.status = ElementStatus::Success;
                                deployment_element.handler_reference = Some(id);
                                database_address
                                    .send(UpdateDeploymentElement(
                                        exercise.id,
                                        deployment_element.clone(),
                                        true,
                                    ))
                                    .await??;

                                if node.type_field == NodeType::VM {
                                    if node.features.is_some() {
                                        (
                                            distributor_address.clone(),
                                            database_address.clone(),
                                            scheduler_address.clone(),
                                            deployers.to_owned(),
                                            self.clone(),
                                            node.clone(),
                                            deployment_element.clone(),
                                            exercise.id,
                                            template_id.clone(),
                                        )
                                            .deploy_features()
                                            .await?;
                                    }
                                    if node.conditions.is_some() {
                                        (
                                            distributor_address.clone(),
                                            database_address.clone(),
                                            deployers.to_owned(),
                                            self.clone(),
                                            node.clone(),
                                            deployment_element,
                                            exercise.id,
                                            template_id,
                                        )
                                            .deploy_conditions()
                                            .await?;
                                    }
                                };

                                Ok(())
                            }
                            Err(error) => {
                                deployment_element.status = ElementStatus::Failed;
                                database_address
                                    .send(UpdateDeploymentElement(
                                        exercise.id,
                                        deployment_element,
                                        true,
                                    ))
                                    .await??;
                                Err(error)
                            }
                        }
                    })
                    .collect::<Vec<_>>(),
            )
            .await?;
        }
        Ok(())
    }
}

#[async_trait]
pub trait RemoveableNodes {
    async fn undeploy_nodes<'a>(
        &'a self,
        distributor_address: &'a Addr<DeployerDistribution>,
        database_address: &'a Addr<Database>,
        deployers: &'a [String],
        exercise_id: &'a Uuid,
    ) -> Result<()>;
}

#[async_trait]
impl RemoveableNodes for Vec<DeploymentElement> {
    async fn undeploy_nodes<'a>(
        &'a self,
        distributor_address: &'a Addr<DeployerDistribution>,
        database_address: &'a Addr<Database>,
        deployers: &'a [String],
        exercise_id: &'a Uuid,
    ) -> Result<()> {
        try_join_all(self.iter().map(|element| async move {
            match element.deployer_type {
                DeployerType(DeployerTypes::VirtualMachine | DeployerTypes::Switch) => {
                    if let Some(handler_reference) = &element.handler_reference {
                        let mut element_update = element.clone();

                        return match distributor_address
                            .send(UnDeploy(
                                element.deployer_type.0,
                                handler_reference.to_string(),
                                deployers.to_owned(),
                            ))
                            .await?
                        {
                            anyhow::Result::Ok(_) => {
                                element_update.status = ElementStatus::Removed;
                                database_address
                                    .send(UpdateDeploymentElement(
                                        exercise_id.to_owned(),
                                        element_update,
                                        true,
                                    ))
                                    .await??;
                                Ok(())
                            }
                            Err(error) => {
                                element_update.status = ElementStatus::RemoveFailed;
                                database_address
                                    .send(UpdateDeploymentElement(
                                        exercise_id.to_owned(),
                                        element_update,
                                        true,
                                    ))
                                    .await??;
                                Err(error)
                            }
                        };
                    }
                    Err(anyhow!("Handler reference not found"))
                }
                _ => Ok(()),
            }
        }))
        .await?;
        Ok(())
    }
}
