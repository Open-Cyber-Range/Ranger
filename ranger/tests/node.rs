mod common;

#[cfg(test)]
mod tests {
    use actix::Actor;
    use actix_rt::System;
    use anyhow::{Error, Result};
    use ranger::node::{CreateNode, DeleteNode, NodeClient};

    #[test]
    fn node_created_successfully() -> Result<()> {
        let socket_address = crate::common::create_mock_node_server().run()?;
        let system = System::new();
        let node_id = system.block_on(async {
            let node_deployer_client =
                NodeClient::new(format!("http://{}", socket_address).clone())
                    .await?
                    .start();
            let node_id = node_deployer_client
                .send(CreateNode(ranger_grpc::Node {
                    name: "some-name".to_string(),
                    exercise_name: "some-exercise-name".to_string(),
                    template_name: "debian10".to_string(),
                }))
                .await??;
            Ok::<String, Error>(node_id)
        })?;

        insta::assert_debug_snapshot!(node_id);
        Ok(())
    }

    #[test]
    fn node_creation_failed() -> Result<()> {
        let socket_address = crate::common::create_mock_node_server()
            .successful_create(false)
            .run()?;
        let system = System::new();
        let result = system.block_on(async {
            let node_deployer_client =
                NodeClient::new(format!("http://{}", socket_address).clone())
                    .await?
                    .start();
            node_deployer_client
                .send(CreateNode(ranger_grpc::Node {
                    name: "some-name".to_string(),
                    exercise_name: "some-exercise-name".to_string(),
                    template_name: "debian10".to_string(),
                }))
                .await??;
            Ok::<(), Error>(())
        });

        assert!(result.is_err());
        insta::assert_debug_snapshot!(result.unwrap_err());
        Ok(())
    }

    #[test]
    fn node_deleted_successfully() -> Result<()> {
        let socket_address = crate::common::create_mock_node_server().run()?;
        let system = System::new();
        system.block_on(async {
            let node_deployer_client =
                NodeClient::new(format!("http://{}", socket_address).clone())
                    .await?
                    .start();
            node_deployer_client
                .send(DeleteNode("some-node-id".to_string()))
                .await??;
            Ok::<(), Error>(())
        })?;

        Ok(())
    }

    #[test]
    fn node_deletion_failed() -> Result<()> {
        let socket_address = crate::common::create_mock_node_server()
            .successful_delete(false)
            .run()?;
        let system = System::new();
        let result = system.block_on(async {
            let node_deployer_client =
                NodeClient::new(format!("http://{}", socket_address).clone())
                    .await?
                    .start();
            node_deployer_client
                .send(DeleteNode("some-node-id".to_string()))
                .await??;
            Ok::<(), Error>(())
        });

        assert!(result.is_err());
        insta::assert_debug_snapshot!(result.unwrap_err());
        Ok(())
    }
}
