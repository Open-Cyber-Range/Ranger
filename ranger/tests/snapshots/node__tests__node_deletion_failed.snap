---
source: ranger/tests/node.rs
expression: result.unwrap_err()
---
"Failed to delete node"
